yfinance/scrapers/fundamentals.py

```diff
@@ -0,0 +1,150 @@
+    def financials(self):
+        if self._financials is None:
+            self._scrape_financials(self.proxy)
+        return self._financials
+
+    @property
+    def shares(self):
+        if self._shares is None:
+            self._scrape_shares(self.proxy)
+        return self._shares
+
+    def _scrape_basics(self, proxy):
+        if self._basics_already_scraped:
+            return
+        self._basics_already_scraped = True
+
+        self._financials_data = self._data.get_json_data_stores(self.ticker_url + '/financials', proxy)
+        try:
+            self._fin_data_quote = self._financials_data['QuoteSummaryStore']
+        except KeyError:
+            err_msg = "No financials data found, symbol may be delisted"
+            print('- %s: %s' % (self._data.ticker, err_msg))
+            return None
+
+    def _scrape_earnings(self, proxy):
+        self._scrape_basics(proxy)
+        # earnings
+        self._earnings = {"yearly": pd.DataFrame(), "quarterly": pd.DataFrame()}
+        if self._fin_data_quote is None:
+            return
+        if isinstance(self._fin_data_quote.get('earnings'), dict):
+            try:
+                earnings = self._fin_data_quote['earnings']['financialsChart']
+                earnings['financialCurrency'] = self._fin_data_quote['earnings'].get('financialCurrency', 'USD')
+                self._earnings['financialCurrency'] = earnings['financialCurrency']
+                df = pd.DataFrame(earnings['yearly']).set_index('date')
+                df.columns = utils.camel2title(df.columns)
+                df.index.name = 'Year'
+                self._earnings['yearly'] = df
+
+                df = pd.DataFrame(earnings['quarterly']).set_index('date')
+                df.columns = utils.camel2title(df.columns)
+                df.index.name = 'Quarter'
+                self._earnings['quarterly'] = df
+            except Exception:
+                pass
+
+    def _scrape_shares(self, proxy):
+        self._scrape_basics(proxy)
+        # shares outstanding
+        try:
+            # keep only years with non None data
+            available_shares = [shares_data for shares_data in
+                                self._financials_data['QuoteTimeSeriesStore']['timeSeries']['annualBasicAverageShares']
+                                if
+                                shares_data]
+            shares = pd.DataFrame(available_shares)
+            shares['Year'] = shares['asOfDate'].agg(lambda x: int(x[:4]))
+            shares.set_index('Year', inplace=True)
+            shares.drop(columns=['dataId', 'asOfDate',
+                                 'periodType', 'currencyCode'], inplace=True)
+            shares.rename(
+                columns={'reportedValue': "BasicShares"}, inplace=True)
+            self._shares = shares
+        except Exception:
+            pass
+
+    def _scrape_financials(self, proxy):
+        self._scrape_basics(proxy)
+        if self._already_scraped_financials:
```

Comment:
Why not just check `if self._financials is not None:` ? Then don't need `_already_scraped_financials`.

Comment also applies to `_basics_already_scraped`

Reply:
I have already rewritten this so financials is fetched in a better way and avoids unnecessary requests.  Can update this draft pull request